csrc/multidevice/utils.cpp

```diff
@@ -97,7 +102,47 @@ bool isResharding(Expr* expr) {
   }
   auto tv_ref = *tvs.begin();
   tvs.erase(tv_ref);
-  return !haveDifferentSharding(tv_ref, tvs).empty();
+  return !getTvsWithDifferentSharding(tv_ref, tvs).empty();
+}
+
+namespace {
+
+void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
+  for (auto tv : tvs) {
+    tv->setDeviceMesh(ref->getDeviceMesh());
+  }
+  scheduler_utils::parallelizeAllLike(ref, tvs, {ParallelType::DIDx});
+}
+
+} // namespace
+
+void insertReshardings(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  for (auto expr : exprs) {
+    if (isLowerableToCommunication(expr)) {
+      continue;
+    }
+    NVF_ERROR(
+        ir_utils::isTvOp(expr),
+        "Non-tv op is not supported yet: ",
+        expr->toString());
+    NVF_ERROR(
+        expr->outputs().size() == 1,
+        "multi-output expressions are not supported");
+    auto output = expr->outputs().at(0)->as<TensorView>();
+    std::vector<TensorView*> new_inputs;
+    for (auto input : getTvsWithDifferentSharding(
+             output, ir_utils::filterByType<TensorView>(expr->inputs()))) {
```

Comment:
I'm feeling a strong uncertainty with this loop. In the loop body, `expr` is being replaced with a new expr that has a new input. Replacing here means the old expr is removed, IIRC, so it seems `expr` in this line becomes a dangling pointer, which should also mean the input vector of the expr becomes invalid.

Reply:
I see. So should I revert back to just producing the container of the TensorView inputs that we iterate over ?

Reply:
Sorry, actually not a concern. `getTvsWithDifferentSharding` is called here, so the iteration should be over the returned set, which shouldn't be affected by the replacement inside the loop body. Sorry for the noise.